app/src/main/java/com/github/multimatum_team/multimatum/util/DeadlineNotification.kt

```diff
@@ -0,0 +1,258 @@
+package com.github.multimatum_team.multimatum.util
+
+import android.app.AlarmManager
+import android.app.NotificationChannel
+import android.app.NotificationManager
+import android.app.PendingIntent
+import android.content.Context
+import android.content.Intent
+import android.content.SharedPreferences
+import android.util.Log
+import androidx.appcompat.app.AppCompatActivity
+import com.github.multimatum_team.multimatum.R
+import com.github.multimatum_team.multimatum.ReminderBroadcastReceiver
+import com.github.multimatum_team.multimatum.model.Deadline
+import com.github.multimatum_team.multimatum.repository.DeadlineID
+import com.github.multimatum_team.multimatum.service.ClockService
+import com.github.multimatum_team.multimatum.service.SystemClockService
+import com.google.gson.Gson
+import dagger.hilt.android.AndroidEntryPoint
+import java.time.ZoneId
+import javax.inject.Inject
+
+
+/**
+ * This service provides function to manage notifications
+ * It provide functions to access and modifiy notifications that are set for a deadline,
+ * A function that verify that notification are bound to existing deadline,
+ * and a function to create notification channel
+ */
+class DeadlineNotification(context: Context) {
+    private lateinit var alarmManager: AlarmManager
+    private lateinit var pendingIntent: PendingIntent
+    private val context= context
+
+    @Inject
+    lateinit var clockService: ClockService
+
+    /**
+     * return a list of all notifications set for this deadline
+     */
+    fun listDeadlineNotification(deadlineId: DeadlineID): List<Long> {
+        val res = readFromSharedPreference(deadlineId)
+        //this is just to remove decimal to the number
+        var resParsed = ArrayList<Long>()
+        for (x in res) {
+             resParsed.add(x.toLong())
+        }
+        return resParsed
+    }
+
+    /**
+     * edit Notification for a given deadline
+     */
+    fun editNotification(deadlineId: DeadlineID, deadline: Deadline, newNotificationTime: List<Long>) {
+        cancelDeadlineNotifications(deadlineId)
+        writeToSharedPreference(deadlineId, newNotificationTime)
+        setDeadlineNotifications(deadlineId, deadline, newNotificationTime)
+        //logNotificationSP()
+    }
+
+    /**
+     * delete all notification for a given deadline
+     */
+    fun deleteNotification(deadlineId: DeadlineID) {
+        cancelDeadlineNotifications(deadlineId)
+        writeToSharedPreference(deadlineId, emptyList())
+        //logNotificationSP()
+    }
+
+    /**
+     * Iterate through existing notification to ensure that there isn't any notification that aren't bound to a deadline.
+     * If such a notification is detected, it'll delete it.
+     */
+    fun updateNotifications(deadlineList:Map<DeadlineID, Deadline>) {
+        val sp = context.getSharedPreferences(
+            context.getString(R.string.notification_shared_preference),
+            Context.MODE_PRIVATE
+        )
```

Comment:
Why don't you use the same `SharedPreferences` instance as the rest of the app? It would be easier and it would allow you to use dependency injection to test this class.

Reply:
Well I think it is more clear to have notification in another file, since there may be many of them. And furthermore to retrieve all notification, we can just call getAll() on this file, so we don't have to filter other value